# CS5062_ASSESSMENT1_ZHIXI_TANG__52097136

Student: ZHIXI TANG
Student ID: 52097136

## Introduction

This .ipynb file include the source code and detailed explanation of CS5062 Assessment 1. It can be run on local machine with
required environment or Google Colab.

**Run on local machine:**

* System:
* Python Version:
* IDE: Pycharm:
* Required packages: see `requirements.txt` or use command `$ pip install -r requirements.txt`

**Run on Colab:**

Simply uploading this .ipynb file on Google Colab then feel free to play it.

## Task 1

### Task 1 - A: Data Import

In [6]:
import numpy.random
import pandas as pd
import numpy as np
import torch

df = pd.read_csv('./data/soybean_tabular.csv')
df

Variety     S_1   S_2     S_3   S_4      M_1   M_2    M_3        W_1  \
0          1  4.0900  15.3  396.90  4.98  0.00632  18.0  6.575   2.381979   
1          2  4.9671  17.8  396.90  9.14  0.02731   0.0  6.421   7.071148   
2          2  4.9671  17.8  392.83  4.03  0.02729   0.0  7.185   6.896941   
3          3  6.0622  18.7  394.63  2.94  0.03237   0.0  6.998   2.237817   
4          3  6.0622  18.7  396.90  5.33  0.06905   0.0  7.147   1.979327   
..       ...     ...   ...     ...   ...      ...   ...    ...        ...   
501        1  2.4786  21.0  391.99  9.67  0.06263   0.0  6.593  11.774062   
502        1  2.2875  21.0  396.90  9.08  0.04527   0.0  6.120  11.785249   
503        1  2.1675  21.0  396.90  5.64  0.06076   0.0  6.976  11.854205   
504        1  2.3889  21.0  393.45  6.48  0.10959   0.0  6.794  11.765196   
505        1  2.5050  21.0  396.90  7.88  0.04741   0.0  6.030  12.007563   

          W_2   W_3         W_4  Yield  
0    0.475522  65.2  296.350195   24.0  
1    0.509165  78.9  241.620198   21.6  
2    0.580673  61.1  241.551476   34.7  
3    0.491539  45.8  222.023994   33.4  
4    0.103660  54.2  221.723972   36.2  
..        ...   ...         ...    ...  
501  0.565469  69.1  272.952382   22.4  
502  0.562102  76.7  273.264306   20.6  
503  0.777569  91.0  273.421726   23.9  
504  0.679057  89.3  273.913622   22.0  
505  0.615699  80.8  273.623418   11.9  

[506 rows x 13 columns]

Provide short statistical summary (mean, range, standard deviations, min/max, median, and 25%/50%/75% percentile),
describe the data provided identifying and rectifying any missing values.

In [7]:
df1 = df.iloc[:, 1:]

# range
df1.describe()

S_1         S_2         S_3         S_4         M_1         M_2  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.795043   18.455534  356.674032   12.653063    3.613524   11.363636   
std      2.105710    2.164946   91.294864    7.141062    8.601545   23.322453   
min      1.129600   12.600000    0.320000    1.730000    0.006320    0.000000   
25%      2.100175   17.400000  375.377500    6.950000    0.082045    0.000000   
50%      3.207450   19.050000  391.440000   11.360000    0.256510    0.000000   
75%      5.188425   20.200000  396.225000   16.955000    3.677083   12.500000   
max     12.126500   22.000000  396.900000   37.970000   88.976200  100.000000   

              M_3         W_1         W_2         W_3         W_4       Yield  
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000  
mean     6.284634   11.131488    0.551258   68.574901  408.252839   22.532806  
std      0.702617    6.868654    0.181693   28.148861  168.530578    9.197104  
min      3.561000    0.472905    0.103660    2.900000  186.765075    5.000000  
25%      5.885500    5.149096    0.422286   45.025000  278.745884   17.025000  
50%      6.208500    9.588040    0.528133   77.500000  330.467783   21.200000  
75%      6.623500   18.094315    0.668573   94.075000  665.354401   25.000000  
max      8.780000   28.074197    1.158538  100.000000  711.210992   50.000000

The statistical description is shown above. As the feature `variety` is just a number to represent corp variety, we do not
have to count its means, std, min, etc.

From the above data we can see in the column `M_2`, there are some values are missed. The dataset has used $0$ to replace
these missed values.

However, to a certain variety of crops, there maybe a certain result corresponding to its features and yield,
therefore, we will get the statistical description of different varieties of corps separately.

In [8]:
# count how many varieties in corps
df['Variety'].value_counts()

24    132
5     115
4     110
3      38
6      26
2      24
8      24
1      20
7      17
Name: Variety, dtype: int64

In [9]:
def show_describe(dataframe, variety_num):
    """"The variety_num can only be one of [24, 5, 4, 3, 6, 2, 8, 1, 7]"""
    df_v = dataframe.loc[dataframe['Variety'] == variety_num]
    return  df_v.describe()

In [10]:
def DiffYield(dataframe):
    varieties = [24, 5, 4, 3, 6, 2, 8, 1, 7]
    index = ['mean', 'std', 'min', '25%', '50%', '75%', 'max']
    v_names = []
    data = {}

    for i in varieties:
        df_v = show_describe(dataframe, i)  # get the describe info of a single variety
        corp_yield = df_v.iloc[1:, -1].values
        exec(f"data['Variety_{i}'] = corp_yield")
    yield_df = pd.DataFrame(data, index=index)

    return yield_df

In [11]:
# df.iloc[1:, -1]
df.iloc[1:, -1].index

RangeIndex(start=1, stop=506, step=1)

In [12]:
DiffYield(df)

Variety_24  Variety_5  Variety_4  Variety_3  Variety_6  Variety_2  \
mean   16.403788  25.706957  21.387273  27.928947  20.976923  26.833333   
std     8.539745   9.328401   6.957883   8.324692   2.312801   7.874376   
min     5.000000  11.800000   7.000000  14.400000  16.800000  15.700000   
25%    11.225000  19.500000  17.575000  21.125000  18.900000  21.400000   
50%    14.400000  23.000000  20.450000  26.500000  21.200000  23.850000   
75%    19.900000  30.000000  23.650000  34.525000  23.025000  33.225000   
max    50.000000  50.000000  50.000000  50.000000  24.800000  43.800000   

      Variety_8  Variety_1  Variety_7  
mean  30.358333  24.365000  27.105882  
std    9.727724   8.024454   6.493215  
min   16.000000  11.900000  17.600000  
25%   23.825000  20.475000  24.300000  
50%   28.250000  22.200000  26.200000  
75%   33.175000  27.225000  29.600000  
max   50.000000  50.000000  42.800000

From above table we can tell that to different varieties of corps, the yield is also different. For example, The Max yield of
`Variety_6` is just 24.8 which is much lower than other varieties. According to the mean of `variety_24`, we can tell its average
yield is much lower than other varieties. Therefore, we can conclude that the variety of corps affects its yield.

### Task 1 - B: Data pre-processing

In this section, we will split dataset to train, validation, test with rate of 6:2:2. At the same time, we will try our
best to ensure the fairness and uniformity of data. According to the conclusion of `Task 1 - A`, we know that the variety
of corps affects its yield. Therefore, we will take samples from each variety at a ratio of 6:2:2 (train:validation:test),
and then compose the datasets. In this case, we can assure as much as possible that in each set of data, the variety proportions of
corps are approximately equal.

In [33]:
from sklearn.preprocessing import StandardScaler

def extract_df_by_variety(dataframe):
    """Split dataframe based on varieties, return all dataframes with dictionary"""
    varieties = [24, 5, 4, 3, 6, 2, 8, 1, 7]
    extracted_df = {}
    for variety in varieties:
        df_s = dataframe.loc[dataframe['Variety'] == variety]
        # exec(f"extracted_df['Variety_{variety}'] = df_s")
        extracted_df[variety] = df_s
    return extracted_df

def split_single_dataset(dataframe):
    """split particular dataset to train:val:test = 6:2:2"""
    train_set = dataframe.sample(frac=0.6, random_state=0, axis=0)
    rest_set = dataframe[~dataframe.index.isin(train_set.index)]
    test_set = rest_set.sample(frac=0.5, random_state=0, axis=0)
    val_set = rest_set[~rest_set.index.isin(test_set.index)]

    return train_set, test_set, val_set

def stratified_sampling(dataframe):
    """"Combine above functions, input the dataframe, return the stratified sampled datasets"""
    extracted_dfs = extract_df_by_variety(dataframe)
    train_sets, test_sets, val_sets = [], [], []
    for _df in extracted_dfs.values():
        train_set, test_set, val_set = split_single_dataset(_df)
        train_sets.append(train_set)
        test_sets.append(test_set)
        val_sets.append(val_set)
    p_train = np.array(pd.concat(train_sets).sample(frac=1), dtype='float32')
    p_test = np.array(pd.concat(test_sets).sample(frac=1), dtype='float32')
    p_val = np.array(pd.concat(val_sets).sample(frac=1), dtype='float32')

    return p_train, p_test, p_val


In [36]:
# get train, validation, test datasets
train_set, test_set, val_set = stratified_sampling(df)


In [37]:
def tensor_generator(dataset):
    """input stratified sampled dataset, return variable x and result y"""
    x = torch.tensor(dataset[...,:12])
    y = torch.tensor(dataset[...,12:])
    return x, y

In [38]:
# get tensors

x_train, y_train = tensor_generator(train_set)
x_val, y_val = tensor_generator(val_set)
x_test, y_test = tensor_generator(test_set)

x_train

tensor([[1.0000e+00, 9.0892e+00, 1.8200e+01,  ..., 2.4773e-01, 3.1500e+01,
         2.4052e+02],
        [4.0000e+00, 3.3175e+00, 1.8400e+01,  ..., 3.5898e-01, 8.2500e+01,
         3.0440e+02],
        [4.0000e+00, 3.6659e+00, 1.8200e+01,  ..., 3.4383e-01, 2.8900e+01,
         2.6969e+02],
        ...,
        [2.4000e+01, 2.5182e+00, 2.0200e+01,  ..., 9.2563e-01, 8.8000e+01,
         6.6515e+02],
        [5.0000e+00, 3.4211e+00, 1.6400e+01,  ..., 5.5014e-01, 8.5100e+01,
         2.7610e+02],
        [2.4000e+01, 2.5052e+00, 2.0200e+01,  ..., 6.8719e-01, 9.1000e+01,
         6.6622e+02]])

According to the first table shown in `Task 1 - A`, we can tell that the values of different features have observable difference, for example,
in the first example, the value of feature `S_3` is 369.90, the value of feature `W_2` is 0.475522. To eliminate the gradient descent majorly
depends on some features, we will do z-score normalization to the variable $X$. As we know, the formula of z-score is: $\hat X = \frac{X-\mu}{\epsilon}$,
where $\hat X$ is the normalized variable $X$, $\mu$ is the mean of $X$, $\epsilon$ is the standard deviation of $X$. Therefore, we can define
function as the following.

In [39]:
# def normalization(dataframe, columns):
#     sc_set = StandardScaler().fit_transform(dataframe)
#     return pd.DataFrame(data=sc_set, columns=columns)
def normalization_x(x, y):
    """Input a tensor, return the z-score normalized tensor"""
    mean = torch.mean(x)
    std = torch.std(x)
    normed_x = (x - mean) / std
    normed_y = (y - mean) / std
    return normed_x, normed_y

# normalize X
normed_x_train, normed_y_train = normalization_x(x_train, y_train)
normed_x_val, normed_y_val = normalization_x(x_val, y_val)
normed_x_test, normed_y_test = normalization_x(x_test, y_test)

In [18]:
from torch.utils import data
class LDataset(data.Dataset):
    def __init__(self, x, y):
        self.x = x
        self.y = y
    def __len__(self):
        return len(self.x)
    def __getitem__(self, idx):
        return self.x[idx], self.y[idx]

train_dataset = LDataset(normed_x_train, normed_y_train)
test_dataset = LDataset(normed_x_test, normed_y_test)
val_dataset = LDataset(normed_x_val, normed_y_val)

test_dataloader = data.DataLoader(test_dataset, batch_size=5, shuffle=True, num_workers=5)
train_dataloader = data.DataLoader(train_dataset, batch_size=10, shuffle=True, num_workers=5)
for i_batch, batched_sample in enumerate(train_dataloader):
    print(f'i_batch: {i_batch}\t-----batched_sample: {batched_sample}')

i_batch: 0	-----batched_sample: [tensor([[-0.4810, -0.4823, -0.3672,  2.0120, -0.3671, -0.4994, -0.5077, -0.4705,
         -0.4529, -0.5049,  0.1487,  1.5498],
        [-0.4810, -0.4679, -0.4007,  2.1209, -0.4583, -0.5074, -0.5077, -0.4639,
         -0.4141, -0.5049, -0.1665,  1.4284],
        [-0.3471, -0.4885, -0.3726, -0.3605, -0.3930, -0.4825, -0.5077, -0.4679,
         -0.3868, -0.5029,  0.0590,  3.9495],
        [-0.4743, -0.4495, -0.3880,  2.1415, -0.4755, -0.5076,  0.0945, -0.4592,
         -0.4994, -0.5059, -0.3612,  1.0049],
        [-0.4810, -0.4841, -0.3846,  2.1367, -0.4384, -0.5050, -0.5077, -0.4650,
         -0.4404, -0.5043, -0.0581,  1.5317],
        [-0.3471, -0.4946, -0.3726, -0.1014, -0.3466, -0.4452, -0.5077, -0.4650,
         -0.3867, -0.5037,  0.1319,  3.9508],
        [-0.4676, -0.4588, -0.3967,  1.9795, -0.4441, -0.5074, -0.1564, -0.4638,
         -0.4723, -0.5040, -0.3545,  1.4564],
        [-0.4810, -0.4813, -0.3846,  2.1448, -0.4308, -0.5060, -0.5077, -0.469

In [41]:
def basic_linear_train(x_train, y_train, x_val, y_val, epoch, lr=0.01, save_model=False, gpu=False, vis=False):
    """The default train function without regularization"""

    model = torch.nn.Linear(12, 1, bias=True)  # define Linear model
    loss_func = torch.nn.MSELoss()  # define loss function
    optim = torch.optim.SGD(model.parameters(), lr=lr, weight_decay=0.001)  # define optimizer

    if gpu:
        model = model.cuda(0)
        x_train = x_train.cuda(0)
        y_train = y_train.cuda(0)
        x_val = x_val.cuda(0)
        y_val = y_val.cuda(0)

    print('iter,\ttrain_loss,\tval_loss')

    for i in range(epoch):
        """Train process"""
        y_hat = model(x_train)
        loss = loss_func(y_hat, y_train)
        optim.zero_grad()
        loss.backward()
        optim.step()
        # print(model.weight.detach().numpy())
        """validation process"""
        y_val_hat = model(x_val)
        loss_val = loss_func(y_val_hat, y_val)
        # print(model.weight.detach().numpy())
        """print the train loss and validation loss"""
        if i % 5000 == 0:
            print(f'{i}\t{loss.item():.4f}\t\t{loss_val.item():.4f}')
    if save_model:
        torch.save(model, './LinearModel.pth')

basic_linear_train(normed_x_train, normed_y_train, normed_x_val, normed_y_val, 100000, 0.001, save_model=True, gpu=True)

iter,	train_loss,	val_loss
0	1.4305		1.3472
5000	0.0031		0.0032
10000	0.0027		0.0029
15000	0.0026		0.0028
20000	0.0024		0.0027
25000	0.0024		0.0027
30000	0.0023		0.0026
35000	0.0023		0.0026
40000	0.0023		0.0026
45000	0.0022		0.0026
50000	0.0022		0.0025
55000	0.0022		0.0025
60000	0.0022		0.0025
65000	0.0022		0.0025
70000	0.0022		0.0025
75000	0.0021		0.0025
80000	0.0021		0.0024
85000	0.0021		0.0024
90000	0.0021		0.0024
95000	0.0021		0.0024


### Task 1 - C: Linear Regression Training

In this section we will define the Linear models and fit them with PyTorch

In [31]:
def basic_linear_train(x_train, y_train, x_val, y_val, epoch, lr=0.01, save_model=False):
    """The default train function without regularization"""

    model = torch.nn.Linear(12, 1, bias=True)  # define Linear model
    loss_func = torch.nn.MSELoss()  # define loss function
    optim = torch.optim.SGD(model.parameters(), lr=lr, weight_decay=0.001)  # define optimizer

    print('iter,\ttrain_loss,\tval_loss')

    for i in range(epoch):
        """Train process"""
        y_hat = model(x_train)
        loss = loss_func(y_hat, y_train)
        optim.zero_grad()
        loss.backward()
        optim.step()
        # print(model.weight.detach().numpy())
        """validation process"""
        y_val_hat = model(x_val)
        loss_val = loss_func(y_val_hat, y_val)
        # print(model.weight.detach().numpy())
        """print the train loss and validation loss"""
        if i % 5000 == 0:
            print(f'{i}\t{loss.item():.4f}\t\t{loss_val.item():.4f}')
    if save_model:
        torch.save(model, './LinearModel.pth')

basic_linear_train(normed_x_train, normed_y_train, normed_x_val, normed_y_val, 100000, 0.001, save_model=True)

iter,	train_loss,	val_loss
0	0.5777		0.5492
5000	0.0026		0.0028
10000	0.0025		0.0028
15000	0.0025		0.0028
20000	0.0024		0.0027
25000	0.0024		0.0027
30000	0.0024		0.0027
35000	0.0024		0.0027
40000	0.0024		0.0027
45000	0.0023		0.0026
50000	0.0023		0.0026
55000	0.0023		0.0026
60000	0.0023		0.0026
65000	0.0023		0.0026
70000	0.0023		0.0026
75000	0.0022		0.0025
80000	0.0022		0.0025
85000	0.0022		0.0025
90000	0.0022		0.0025
95000	0.0022		0.0025


In [43]:
def test_data(model_path, x_test, y_test, visualize=False):
    model = torch.load(model_path)
    loss_func = torch.nn.MSELoss()
    predict = model(x_test)
    loss = loss_func(y_test, predict)
    print(f'test loss: {loss:.4f}')

test_data('./LinearModel.pth', normed_x_test.cuda(0), normed_y_test.cuda(0))

test loss: 0.0019


In [45]:
def linear_L1(x_train, y_train, x_val, y_val, epoch, lr=0.001, save_model=False, gpu=False, vis=False):

    model = torch.nn.Linear(12, 1, bias=True)  # define Linear model
    loss_func = torch.nn.MSELoss()  # define loss function
    optim = torch.optim.SGD(model.parameters(), lr=lr)  # define optimizer

    if gpu:
        model = model.cuda(0)
        x_train = x_train.cuda(0)
        y_train = y_train.cuda(0)
        x_val = x_val.cuda(0)
        y_val = y_val.cuda(0)
    print('iter,\ttrain_loss,\tval_loss')

    for i in range(epoch):
        """Train process"""
        w = 0
        for param in model.parameters():
            w += torch.sum(abs(param))
        y_hat = model(x_train)
        loss = loss_func(y_hat, y_train) + 0.001 * w
        optim.zero_grad()
        loss.backward()
        optim.step()
        # print(model.weight.detach().numpy())
        """validation process"""
        y_val_hat = model(x_val)
        loss_val = loss_func(y_val_hat, y_val)
        # print(model.weight.detach().numpy())
        """print the train loss and validation loss"""
        if i % 5000 == 0:
            print(f'{i}\t{loss.item():.4f}\t\t{loss_val.item():.4f}')
    if save_model:
        torch.save(model, './LinearModel_L1.pth')

linear_L1(normed_x_train, normed_y_train, normed_x_val, normed_y_val, 100000, lr=0.001, save_model=True, gpu=True)
test_data('./LinearModel_L1.pth', normed_x_test.cuda(0), normed_y_test.cuda(0))

# TODO: Revise test_data function, refer the materials, re-write mse, rmse, mae. Implement them for q-3.
# TODO: Try to draw the graph for observation. 
# TODO: Finish the report part, start task 2. 

iter,	train_loss,	val_loss
0	0.0370		0.0402
5000	0.0101		0.0109
10000	0.0061		0.0060
15000	0.0050		0.0047
20000	0.0047		0.0043
25000	0.0045		0.0041
30000	0.0043		0.0039
35000	0.0041		0.0038
40000	0.0040		0.0037
45000	0.0039		0.0036
50000	0.0038		0.0036
55000	0.0037		0.0035
60000	0.0037		0.0035
65000	0.0036		0.0035
70000	0.0035		0.0034
75000	0.0035		0.0034
80000	0.0035		0.0034
85000	0.0034		0.0034
90000	0.0034		0.0034
95000	0.0034		0.0033
test loss: 0.0027


In [18]:
import torch
import matplotlib.pyplot as plt
import numpy as np